### ISSUE TITLE
nice_fit() error when estimating lavaan via tidySEM

### ISSUE THREAD

**Issue by AlexGareau** [rempsyc/lavaanExtra#7]
the function nice_fit() works fine when estimating directly with lavaan a cfa() or sem().
However, while using tidySEM to estimate_lavaan(), which produce a lavaan object, nice_fit() return this error.

```
Error in vapply(match.call(expand.dots = FALSE)$..., as.character, FUN.VALUE = "character") : 
  values must be length 1,
 but FUN(X[[1]]) result is length 3
```

I wonder where the issue is, because I was able to reproduce the lavaan fit object with pure-lavaan and tidySEM.

**Comment by rempsyc** [comment]
Interesting, thanks for raising this issue! I will look into it and come back to you.

**Comment by rempsyc** [comment]
I just made a very simple reprex below. A reprex is a minimally reproducible example that helps us identify the source of the bug. In this case, I use the basic example given in `estimate_lavaan()` and feed it to `nice_fit` to see if this basic example works. Let's have a look at the results:

``` r
library(lavaan)
library(lavaanExtra)
library(tidySEM)

model <- tidy_sem(iris, "\\.")
model <- measurement(model)
res <- estimate_lavaan(model)
#> Warning in lav_object_post_check(object): lavaan WARNING: some estimated ov
#> variances are negative
#> Warning in lav_object_post_check(object): lavaan WARNING: covariance matrix of latent variables
#>                 is not positive definite;
#>                 use lavInspect(fit, "cov.lv") to investigate.

nice_fit(res)
#>   Model  chi2 df chi2.df     p   CFI   TLI RMSEA  SRMR     AIC     BIC
#> 1   res 2.322  1   2.322 0.128 0.998 0.989 0.094 0.014 788.151 827.289
```

``` r
nice_fit(res, res)
#>   Model  chi2 df chi2.df     p   CFI   TLI RMSEA  SRMR     AIC     BIC
#> 1   res 2.322  1   2.322 0.128 0.998 0.989 0.094 0.014 788.151 827.289
#> 2   res 2.322  1   2.322 0.128 0.998 0.989 0.094 0.014 788.151 827.289
```

<sup>Created on 2022-11-10 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Good news, seems like it works! Bad news is that could mean it is something specific with your data set. Would it be possible for you to attempt running the above code see if at least this works for you?

If the above works for you, then it would be helpful to try to identify when the bug arises and when it does not. The best scenario would be if you could provide an example data set that could reproduce this bug. This would make it easier for me to correct the bug.

I wonder for instance if it would be possible that you misspecified the `tidy_sem` syntax (e.g., with "\\."). What happens when you use summary on the tidy_sem object you've created? Do you get the regular `lavaan` output?

Once again, thank you for posting here and using this package!

**Comment by rempsyc** [comment]
Nice to hear! I'm not sure why `tidy_sem` asks for "\\."), but from the documentation, it seems to refer to the `split` argument, which uses regex notation:

```r
split	
Character. Defining the regular expression used by strsplit to separate variable names into 1) the name of the 
scale/construct and 2) the number (or name) of the item.
```

The first backslash `\` escapes the next `\` (else it would create an error since it is a special character), the second `\` escapes the dot `.` (also a special character), and based on regex, I believe the dot `.` means "use everything". Quite complicated for what it is, but that is regex for you! :)

**Comment by AlexGareau** [comment]
Well, unfortunately, I said "it work" too fast. I probably had a lavaan object loaded in environment with the same name when I tested your solution `tidy_sem("\\.")`. I can't found the documentation you are referring to for the "\.".

Anyhow, I might have forget to tell you that the lavaan object comes from a list of lavaan objects that have been created with `lapply()`

I tested some alternative code and I think it comes from the fact that I am extracting from a list. Again here is the error...

```
Error in vapply(match.call(expand.dots = FALSE)$..., as.character, FUN.VALUE = "character") : 
  values must be length 1,
 but FUN(X[[1]]) result is length 3
```
And here is some codes that work if you assign the lavaan object to a seperate object instead of going through the list.
```
#produce the errors
nice_fit(cfa.fit[[1]])
nice_fit(cfa.fit$wave2)
nice_fit(cfa.fit[1])

#no errors this way
test <- cfa.fit[[1]]
test <- cfa.fit$wave2
nice_fit(test)
```

**Comment by rempsyc** [comment]
Here is a reprex summarizing the issue:

``` r
library(lavaan)
library(lavaanExtra)
library(tidySEM)

model <- tidy_sem(iris, "\\.")
model <- measurement(model)
res <- estimate_lavaan(model)

# Create a list of lavaan objects
res.list <- list(res, res, res)
names(res.list) <- paste0("Mod", 1:3)

#produce the errors
nice_fit(res.list[[1]])
#> Error in vapply(match.call(expand.dots = FALSE)$..., as.character, FUN.VALUE = "character"): values must be length 1,
#>  but FUN(X[[1]]) result is length 3
nice_fit(res.list$Mod1)
#> Error in vapply(match.call(expand.dots = FALSE)$..., as.character, FUN.VALUE = "character"): values must be length 1,
#>  but FUN(X[[1]]) result is length 3

#no errors this way
test <- res.list[[1]]
nice_fit(test)
#>   Model  chi2 df chi2.df     p   CFI   TLI RMSEA  SRMR     AIC     BIC
#> 1  test 2.322  1   2.322 0.128 0.998 0.989 0.094 0.014 788.151 827.289

test <- res.list$Mod1
nice_fit(test)
#>   Model  chi2 df chi2.df     p   CFI   TLI RMSEA  SRMR     AIC     BIC
#> 1  test 2.322  1   2.322 0.128 0.998 0.989 0.094 0.014 788.151 827.289

#this works too
lapply(res.list,function(x){
  nice_fit(x)
})
#> $Mod1
#>   Model  chi2 df chi2.df     p   CFI   TLI RMSEA  SRMR     AIC     BIC
#> 1     x 2.322  1   2.322 0.128 0.998 0.989 0.094 0.014 788.151 827.289
#> 
#> $Mod2
#>   Model  chi2 df chi2.df     p   CFI   TLI RMSEA  SRMR     AIC     BIC
#> 1     x 2.322  1   2.322 0.128 0.998 0.989 0.094 0.014 788.151 827.289
#> 
#> $Mod3
#>   Model  chi2 df chi2.df     p   CFI   TLI RMSEA  SRMR     AIC     BIC
#> 1     x 2.322  1   2.322 0.128 0.998 0.989 0.094 0.014 788.151 827.289

# Compare with performance::compare_performance that uses the same strategy
performance::compare_performance(res, res)
#> Warning: When comparing models, please note that probably not all models were fit
#>   from same data.
#> # Comparison of Model Performance Indices
#> 
#> Name |  Model | Chi2(1) | p (Chi2) | Baseline(6) | p (Baseline) |   GFI |  AGFI |   NFI |  NNFI |   CFI | RMSEA |    RMSEA  CI | p (RMSEA) |   RMR |  SRMR |   RFI |  PNFI |   IFI |   RNI | Loglikelihood | AIC (weights) | BIC (weights) | BIC_adjusted
#> ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
#> res  | lavaan |   2.322 |    0.128 |     722.206 |       < .001 | 1.000 | 0.998 | 0.997 | 0.989 | 0.998 | 0.094 | [0.00, 0.26] |     0.197 | 0.006 | 0.014 | 0.981 | 0.166 | 0.998 | 0.998 |      -381.076 | 788.2 (>.999) | 827.3 (>.999) |      786.147

# This should theoretically not work
performance::compare_performance(res.list[[1]])
#> Error in as.list.default(X): no method for coercing this S4 class to a vector
performance::compare_performance(res.list$Mod1)
#> Error in as.list.default(X): no method for coercing this S4 class to a vector

# Just to confirm that this is the same without involving tidySEM
HS.model <- ' visual  =~ x1 + x2 + x3
              textual =~ x4 + x5 + x6
              speed   =~ x7 + x8 + x9 '

res <- cfa(HS.model, data = HolzingerSwineford1939)

# Create a list of lavaan objects
res.list <- list(res, res, res)
names(res.list) <- paste0("Mod", 1:3)

#produce the errors
nice_fit(res.list[[1]])
#> Error in vapply(match.call(expand.dots = FALSE)$..., as.character, FUN.VALUE = "character"): values must be length 1,
#>  but FUN(X[[1]]) result is length 3
nice_fit(res.list$Mod1)
#> Error in vapply(match.call(expand.dots = FALSE)$..., as.character, FUN.VALUE = "character"): values must be length 1,
#>  but FUN(X[[1]]) result is length 3

#no errors this way
test <- res.list[[1]]
nice_fit(test)
#>   Model   chi2 df chi2.df p   CFI   TLI RMSEA  SRMR     AIC      BIC
#> 1  test 85.306 24   3.554 0 0.931 0.896 0.092 0.065 7517.49 7595.339

test <- res.list$Mod1
nice_fit(test)
#>   Model   chi2 df chi2.df p   CFI   TLI RMSEA  SRMR     AIC      BIC
#> 1  test 85.306 24   3.554 0 0.931 0.896 0.092 0.065 7517.49 7595.339

#this works too
lapply(res.list,function(x){
  nice_fit(x)
})
#> $Mod1
#>   Model   chi2 df chi2.df p   CFI   TLI RMSEA  SRMR     AIC      BIC
#> 1     x 85.306 24   3.554 0 0.931 0.896 0.092 0.065 7517.49 7595.339
#> 
#> $Mod2
#>   Model   chi2 df chi2.df p   CFI   TLI RMSEA  SRMR     AIC      BIC
#> 1     x 85.306 24   3.554 0 0.931 0.896 0.092 0.065 7517.49 7595.339
#> 
#> $Mod3
#>   Model   chi2 df chi2.df p   CFI   TLI RMSEA  SRMR     AIC      BIC
#> 1     x 85.306 24   3.554 0 0.931 0.896 0.092 0.065 7517.49 7595.339

# Compare with performance::compare_performance that uses the same strategy
performance::compare_performance(res, res)
#> Warning: When comparing models, please note that probably not all models were fit
#>   from same data.
#> # Comparison of Model Performance Indices
#> 
#> Name |  Model | Chi2(24) | p (Chi2) | Baseline(36) | p (Baseline) |   GFI |  AGFI |   NFI |  NNFI |   CFI | RMSEA |    RMSEA  CI | p (RMSEA) |   RMR |  SRMR |   RFI |  PNFI |   IFI |   RNI | Loglikelihood |  AIC (weights) |  BIC (weights) | BIC_adjusted
#> -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
#> res  | lavaan |   85.306 |   < .001 |      918.852 |       < .001 | 0.943 | 0.894 | 0.907 | 0.896 | 0.931 | 0.092 | [0.07, 0.11] |    < .001 | 0.082 | 0.065 | 0.861 | 0.605 | 0.931 | 0.931 |     -3737.745 | 7517.5 (>.999) | 7595.3 (>.999) |     7528.739

# This should theoretically not work
performance::compare_performance(res.list[[1]])
#> Error in as.list.default(X): no method for coercing this S4 class to a vector
performance::compare_performance(res.list$Mod1)
#> Error in as.list.default(X): no method for coercing this S4 class to a vector

# Workaround 1
lapply(res.list,function(x){
  nice_fit(x)
}) |> dplyr::bind_rows()
#>   Model   chi2 df chi2.df p   CFI   TLI RMSEA  SRMR     AIC      BIC
#> 1     x 85.306 24   3.554 0 0.931 0.896 0.092 0.065 7517.49 7595.339
#> 2     x 85.306 24   3.554 0 0.931 0.896 0.092 0.065 7517.49 7595.339
#> 3     x 85.306 24   3.554 0 0.931 0.896 0.092 0.065 7517.49 7595.339
# But then the model names are all named x, so that's not ideal...

# Workaround 2, like you suggested:
mod1 <- res.list[[1]]
mod2 <- res.list[[2]] 
mod3 <- res.list[[3]] 

nice_fit(mod1, mod2, mod3)
#>   Model   chi2 df chi2.df p   CFI   TLI RMSEA  SRMR     AIC      BIC
#> 1  mod1 85.306 24   3.554 0 0.931 0.896 0.092 0.065 7517.49 7595.339
#> 2  mod2 85.306 24   3.554 0 0.931 0.896 0.092 0.065 7517.49 7595.339
#> 3  mod3 85.306 24   3.554 0 0.931 0.896 0.092 0.065 7517.49 7595.339
```

<sup>Created on 2022-11-18 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by rempsyc** [comment]
So for now I would suggest using the workaround 2 from my example above (actually your initial workaround).

That said, I am still experimenting with this unconventional method of using the `...` argument to avoid using a list object.

What's your preference? (1) feeding the models directly to the function without having to define it as list (but does not work with providing a list), or (2) force having to provide a list (but will always need to provide a list when having more than 2 models, instead of the models directly)?

**Comment by AlexGareau** [comment]
Thank you for putting such effort in documenting this little issue. 
My preferences would be to be with a list feature, because I think that most of large workflow uses list to store iteration. In a SEM framework, researchers are making different type of comparison across models that either within the same sample (with constraint) or by comparing an equivalent model across sample. Working with list seems like a convenient way of programmatically do those comparison... I might be wrong; I only have been working with R for 1-year.

**Comment by rempsyc** [comment]
Thanks for your patience @AlexGareau! This has (finally) been fixed in cc163bcb9e7c5fdec39e32b2ec2b7d9aa5f8e2d2 and the new version available on GitHub which I am about to push to CRAN. You can now feed `nice_fit` objects of class `list`. Using a named list will use the list names to populate the model names in the table. You can name your list automatically using `tibble::lst()` instead of base R `list()` (I just discovered this and it is amazing. It always bothered me having to name my lists manually).
